File: solutions/fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One = 1
    | Two = 2
    | Three = 3
    | Four = 4
    | Five = 5
    | Six = 6


let private sumForValue (value: int) (dice: int list): int =
    dice |> List.filter(fun die -> die = value) |> List.sum

let private checkYacht (dice: int list): int =
    if dice |> List.distinct |> List.length = 1 then 50 else 0

let private checkStraight (values: int list) (dice: int list): int =
    if dice |> List.sort = values then 30 else 0

let private checkFourOfAKind (dice: int list): int =
    match dice |> List.sort with
    | [d1; d2; d3; d4; d5] -> if (d1 = d4 || d2 = d5) then d3 * 4 else 0
    | _ -> 0

let private checkFullHouse (dice: int list): int =
    match dice |> List.countBy id with
    | [(_, 2); (_, 3)] -> dice |> List.sum
    | _ -> 0


let score (category: Category) (dice: Die list) = 
    dice |> List.map int |>
    match category with
    | Ones           -> sumForValue 1
    | Twos           -> sumForValue 2
    | Threes         -> sumForValue 3
    | Fours          -> sumForValue 4
    | Fives          -> sumForValue 5
    | Sixes          -> sumForValue 6
    | FullHouse      -> checkFullHouse
    | FourOfAKind    -> checkFourOfAKind
    | LittleStraight -> checkStraight [1; 2; 3; 4; 5]
    | BigStraight    -> checkStraight [2; 3; 4; 5; 6]
    | Yacht          -> checkYacht
    | Choice         -> List.sum

File: solutions/csharp/yacht/Yacht.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{
    public static int Score(int[] dice, YachtCategory category) => category switch
    {
        YachtCategory.Ones => sumForValue(dice, 1),
        YachtCategory.Twos => sumForValue(dice, 2),
        YachtCategory.Threes => sumForValue(dice, 3),
        YachtCategory.Fours => sumForValue(dice, 4),
        YachtCategory.Fives => sumForValue(dice, 5),
        YachtCategory.Sixes => sumForValue(dice, 6),

        YachtCategory.FullHouse => checkFullHouse(dice),
        YachtCategory.FourOfAKind => checkFourOfAKind(dice),
        YachtCategory.LittleStraight => checkStraight(dice, 1),
        YachtCategory.BigStraight => checkStraight(dice, 2),
        YachtCategory.Yacht => checkYacht(dice),
        YachtCategory.Choice => dice.Sum(),

        _ => 0,
    };

    private static int checkYacht(int[] dice) =>
        dice.Distinct().Count() == 1 ? 50 : 0;

    private static int checkFourOfAKind(int[] dice)
    {
        var groupOfMostDice = dice.GroupBy(die => die).largestGroup();
        return groupOfMostDice.Count() >= 4 ? groupOfMostDice.Key * 4 : 0;
    }

    private static int checkFullHouse(int[] dice)
    {
        var groupsOfDice = dice.GroupBy(die => die);
        if (groupsOfDice.Count() != 2) return 0;
        return groupsOfDice.largestGroup().Count() == 3 ? dice.Sum() : 0;
    }

    private static int checkStraight(int[] dice, int start) =>
        dice.OrderBy(die => die).SequenceEqual(Enumerable.Range(start, 5)) ? 30 : 0;

    private static int sumForValue(int[] dice, int value) =>
        dice.Where(die => die == value).Sum();

    private static IGrouping<T, T> largestGroup<T>(this IEnumerable<IGrouping<T, T>> groups) =>
        groups.OrderByDescending(g => g.Count()).First();
}

